crates/cext/src/circuit.rs

```diff
@@ -682,3 +683,112 @@ pub unsafe extern "C" fn qk_circuit_delay(
 
     ExitCode::Success
 }
+
+/// @ingroup QkCircuit
+/// Compose a circuit into other circuit
+///
+/// @param target A pointer to the circuit to compose a circuit
+/// @param other A pointer to the circuit to be composed into
+/// @param qubits A pointer to array of qubits used to map other circuit to target
+/// @param clbits A pointer to array of clbits used to map other circuit to target
+/// @param front insert other circuit onto the top of target circuit
+/// @param inplace insert other circuit into target itself, returns null pointer
+///        both inplace and front can not be set to true at the same time
+///
+/// @return a new composed circuit, if inplace = true returns null pointer
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     QkCircuit *qc_sub = qk_circuit_new(2);
+///     qk_circuit_compose(qc, qc_sub, *[3, 2], *[3, 2], false, false);
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_compose(
+    target: *mut CircuitData,
+    other: *const CircuitData,
+    qubits: *const u32,
+    clbits: *const u32,
+    front: bool,
+    inplace: bool,
+) -> *mut CircuitData {
+    let target = unsafe { mut_ptr_as_ref(target) };
+    let other = unsafe { const_ptr_as_ref(other) };
+    let num_qubits = other.num_qubits();
+    let num_clbits = other.num_clbits();
+    let qubits: Vec<u32> = unsafe {
+        (0..num_qubits)
+            .map(|idx| *qubits.wrapping_add(idx))
+            .collect()
+    };
```

Comment:
We immediately dereference `qubits.wrapping_add(idx)` without further calculation, so there's no need to use `wrapping_add`; we can just use the regular `add` because it'd be UB if there was overflow anyway.